### ISSUE TITLE
DINOv3から抽出した特徴量をボール、コート、プレーヤー(ポーズ)の軽量なデコーダに渡し推論させる

### ISSUE THREAD

**Issue by Motoki0705** [Motoki0705/3D-Tennis-Lab#28]
# dino-sequential プロジェクト仕様（v0.2）

## 1. 目的

* エンコーダを **DINOv3 (ViT-S)** に統一し、ボール／プレーヤー／コートの複数タスクで共有して計算量を削減。
* **時系列抽出器（ConvGRU）** をボール・プレーヤーで共有し、マルチモーダル化を検証。
* タスク学習の配分・スケジューリングを設計し、効率的なマルチタスク学習を実現。

---

## 2. タスク定義

* **Ball**：キーポイント抽出（出力は **heatmap**）
* **Player**：検出（まず **Faster R-CNN**、次に **DETR decoder** を試行）。将来的にグローバル特徴から **pose 推定**へ拡張の可能性。
* **Court**：キーポイント抽出（**heatmap**、時系列不要・独立学習。※既学習済み）

---

## 3. 入出力フォーマット

* **入力（映像）**：`[B, N, C, H, W]`（B: バッチ, N: 時系列長）
* **アノテーション**：各フレームのキーポイント座標 (x, y)
* **学習ターゲット**：

  * Ball／Court：heatmap（座標をガウシアンに展開して教師化）
  * Player：検出ラベル（Faster/DETR 用）

**出力解像度**：`H/4 × W/4`（heatmap 出力）

**σ の決め方**：**スケール依存**（画像スケール／バウンディング領域などに応じて動的決定）。

---

## 4. モデル構成

```
Frames [B,N,C,H,W]
   └─ DINOv3 (ViT-S)  ← 共有エンコーダ（空間特徴）
        ├─(Ball/Player)→ ConvGRU（時系列抽出器, 共有, 初期=恒等近傍）
        │                 └→ Task Decoders
        │                     - Ball: Heatmap Head
        │                     - Player: Faster R-CNN or DETR decoder
        └─(Court)────────────→ Court Heatmap Head（時系列スキップ, 独立学習済）
```

### 4.1 DINOv3 (ViT-S)

* 出力例：`[B, N, P, D]`（P: パッチ数, D: 次元）または `[B,N,Ch,Hp,Wp]`（再配置）
* **学習スキーム**：**完全凍結（full freeze）**

### 4.2 時系列抽出器（ConvGRU）

* **時系列長**：固定／可変の両対応
* **設定**：**コンフィグベース**（層数、隠れチャネル、カーネルサイズ、正規化などを YAML/JSON で指定）
* **初期化**：恒等写像近傍（ゲート初期値・アダプタConvで安定化）
* **学習率**：デコーダより低く設定（例：`lr_seq = lr_dec * α`, α∈{0.1, 0.01}）

### 4.3 デコーダ

* **Ball/Court Heatmap Head**：Conv→Upsample→Conv…→`K` チャンネル（K: キーポイント数。Ball は通常 1）
* **Player**：

  * **Faster R-CNN**：ConvGRU 後の特徴マップをバックボーン相当として利用（チャネル合わせのアダプタConv）
  * **DETR decoder**：ConvGRU 後特徴（flatten+pos enc）→ Transformer decoder

---

## 5. データセット

* `core/BaseSequenceDataset` を継承

  * 入力：共通フレーム `[B,N,C,H,W]`
  * ターゲット：

    * Ball：各フレームの (x, y) → heatmap 化
    * Player：各フレームの bbox/label（DETR なら set-based 形式に変換）
* 変換／前処理：

  * 画像増強（flip, color jitter, random crop/resize）
  * 座標・bbox は幾何変換に追従
  * heatmap 生成時の σ はスケールに応じて動的決定

---

## 6. 損失と最適化

* **総損失**：`L_total = w_ball * L_ball + w_player * L_player`（**w\_ball = w\_player = 1** を基本値として実装で重み付け可能）
* **Ball/Court（heatmap）**：MSE または Focal-heatmap（スパース性が高い場合に有効）。必要に応じて soft-argmax 補助損失。
* **Player（検出）**：

  * Faster：分類 CE + 回帰 L1/GIoU
  * DETR：Hungarian matching + クラス CE + L1 + GIoU
* **学習率**：

  * デコーダ：基準学習率 `lr_dec`
  * ConvGRU：`lr_dec * α`（例：α=0.1〜0.01）
  * DINOv3：**凍結**（finetune なし）

---

## 7. 学習スケジュールとロギング

* **同時学習**：Ball と Player を同一バッチ（同一入力）で同時最適化
* **Court**：時系列スキップで**独立**（既学習済み）
* **TensorBoard 出力**：

  * Heatmap（GT/Pred のオーバーレイ）
  * Player 検出の可視化（bbox & score）
  * 損失曲線と学習率スケジューラの可視化

---

## 8. 設定例（YAML スケルトン）

```yaml
model:
  encoder: dinov3_vit_s
  encoder_freeze: true
  seq:
    type: convgru
    cfg: ${convgru_cfg}  # 層数/Ch/Kernel/Norm などをここで指定
  heads:
    ball: heatmap
    player: faster_rcnn   # or detr
    court: heatmap        # offline / pre-trained
loss:
  ball: mse               # or focal_heatmap
  player: faster_default  # or detr_default
  weights:
    ball: 1.0
    player: 1.0
optim:
  lr_decoder: 1.0e-3
  lr_seq_scale: 0.1       # seq lr = lr_decoder * 0.1
  weight_decay: 1.0e-4
  scheduler: cosine       # 任意
  grad_clip: 1.0          # 任意
data:
  input_shape: [B,N,C,H,W]
  heatmap:
    out_stride: 4         # H/4, W/4
    sigma: scale_dependent
log:
  tensorboard: true
```

---

## 9. リスクと対策（要点）

* **ボールの極端なスパース** → focal-heatmap／adaptive σ／hard negative mining
* **ConvGRU の不安定** → 残差・正規化・短い BPTT、学習率低減
* **検出器との特徴整合** → アダプタ層（Conv/Norm）、位置埋め込みの整備（DETR）
* **タスク干渉** → 勾配正規化（GradNorm/PCGrad）をオプション化

---

## 10. 未確定事項（要ヒアリング）

1. **σ のスケール依存の具体式**

   * 例：`σ = k * max(h_bbox, w_bbox)` または 画像短辺比率など。`k` の標準値を設定したい。
2. **ConvGRU のコンフィグ項目**

   * `num_layers, hidden_channels, kernel_size, norm(=BN/IN/GN), dropout` などの初期値。
3. **Player 検出ラベル仕様**

   * bbox は人領域のみ／クラスは単一か複数か。ID 追跡は今回対象外で良いか。
4. **DETR 側の基本設定**

   * `num_queries`、positional encoding（2D/learned/sine）、decoder layers の深さ。
5. **訓練の実務パラメータ**

   * バッチサイズ、時系列長 N、解像度、オプティマイザ（AdamW 予定？）、学習エポック数。

---

### 付録：heatmap 生成の実装メモ

* 出力解像度が H/4×W/4 なので、座標も 1/4 スケールへ射影。
* ピクセル中心でのガウシアン核を生成（半径は 3σ 程度をカバー）。
* ミニバッチで可変 N に対応する場合、padding と mask を併用（損失計算でマスク）。
